### ISSUE TITLE
investigate Haskell’s suitability

### ISSUE THREAD

**Issue by mhuesch** [sacredcapital/ReputationDSL#3]
## N.B.

this is further-future, not something to act on yet.

## synopsis

right now, (I believe) Holochain only has an HDK for Rust.

it seems possible that Haskell might be useful to us, because of easy access to things like type classes and higher-order functional constructs. instead of having to implement a programming language in Rust, it’s possible we could do an embedded DSL in Haskell, which would allow us to use Haskell’s type system and programming constructs (e.g. monads) as a “shallow embedding”.

the idea here is to scope the work required to use Haskell for the reputation DSL.

## additional motivation that is pretty subjective/personal

I @mhuesch am better with Haskell than Rust, so dev work might go significantly faster 🙂

### Haskell interpreter / embedded DSL architecture

- the need:
    - we need something which can be compiled to WASM, which can interpret user-provided programs.
    - we would like to have higher-order constructs & perhaps things like monads
    - why not use Haskell itself?
    - thus we would use [`hint`](https://hackage.haskell.org/package/hint) to interpret Haskell code. we would compile a WASM binary which contained `hint` and thus could interpret Haskell code inside of Holochain.
- as I demonstrated [here](https://github.com/mhuesch/hint-playground/blob/master/src/MyLib.hs#L33-L40) we could use `hint` to load Haskell modules and then interpret their contents.
    - this would allow us to load Haskell modules out of a Holochain DHT, then pass them to `hint` to parse/typecheck/interpret.
- the interface might look like:
    - we (authors of DSL) define machinery which handles validation, type marshalling, interoperation with Reputation Vault, etc.
    - users would provide reputation computation programs, as Haskell modules which define certain specially-named functions (we, authors of DSL, would decide these names and communicate them).
    - `hint` would interpret user-provided modules, in conjunction with DSL-core modules, in order to create & execute the reputation programs with "live" Holochain data.

### WASM compilation of Haskell

Tweag have [asterius](https://github.com/tweag/asterius) which compiles Haskell -> WASM. it's a modified version of GHC. I have not yet played with it much, besides failing to build [this branch](https://github.com/tweag/asterius/tree/david-nix) on NixOS.

### Holochain HDK implementation (ABI compatibility with Holochain)

this is maybe the biggest question mark? Holochain has been moving fast and I'm not sure of stability of APIs with RSM.

we would need to write an HDK for Haskell, which would require some fairly understanding of Holochain's core and how to interface between zomes written in different languages.

## overall

to me the question is whether the cost of supporting Haskell on Holochain (building HDK, getting WASM compilation working) is worth the payoff of having Haskell's power. I'm not sure what the answer is.

**Comment by mhuesch** [comment]
something slightly unfortunate: `hint` seems to make the binary quite large:

67M, which reduces to 55M with `strip`.

maybe this would be a good stress-test for Holochain's WASM engine 🤠

```
vayu :: ~/projects/hint-playground ‹master*› » cabal v2-build
Resolving dependencies...
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - hint-playground-0.1.0.0 (lib) (configuration changed)
 - hint-playground-0.1.0.0 (exe:hp) (configuration changed)
Configuring library for hint-playground-0.1.0.0..
Preprocessing library for hint-playground-0.1.0.0..
Building library for hint-playground-0.1.0.0..
Configuring executable 'hp' for hint-playground-0.1.0.0..
Preprocessing executable 'hp' for hint-playground-0.1.0.0..
Building executable 'hp' for hint-playground-0.1.0.0..
[1 of 1] Compiling Main             ( app/Main.hs, /home/mhueschen/projects/hint-playground/dist-newstyle/build/x86_64-linux/ghc-8.6.5/hint-playground-0.1.0.0/x/hp/build/hp/hp-tmp/Main.o ) [Optimisation flags changed]
Linking /home/mhueschen/projects/hint-playground/dist-newstyle/build/x86_64-linux/ghc-8.6.5/hint-playground-0.1.0.0/x/hp/build/hp/hp ...
vayu :: ~/projects/hint-playground ‹master*› » ls -lh ./dist-newstyle/build/x86_64-linux/ghc-8.6.5/hint-test-0.1.0.0/x/hint-test/build/hint-test/hint-test
-rwxr-xr-x 1 mhueschen users 67M Dec 14 14:07 ./dist-newstyle/build/x86_64-linux/ghc-8.6.5/hint-test-0.1.0.0/x/hint-test/build/hint-test/hint-test
vayu :: ~/projects/hint-playground ‹master*› » strip ./dist-newstyle/build/x86_64-linux/ghc-8.6.5/hint-test-0.1.0.0/x/hint-test/build/hint-test/hint-test
vayu :: ~/projects/hint-playground ‹master*› » ls -lh ./dist-newstyle/build/x86_64-linux/ghc-8.6.5/hint-test-0.1.0.0/x/hint-test/build/hint-test/hint-test
-rwxr-xr-x 1 mhueschen users 55M Dec 14 14:55 ./dist-newstyle/build/x86_64-linux/ghc-8.6.5/hint-test-0.1.0.0/x/hint-test/build/hint-test/hint-test
```

**Comment by mhuesch** [comment]
of interest to HDK development, as types are defined here: https://github.com/holochain/holochain/tree/develop/crates/holochain_zome_types.

**Comment by mhuesch** [comment]
so today I tried the experiment of running `asterius` on my `hint-playground` package, attempting to interpret a Haskell source file in a wasm-compiled version of `hint` (the Haskell interpreter). I expected this to fail, and I ran into a roadblock that seems failure-like, but I am not totally sure how to interpret it yet.

I documented it [here](https://github.com/mhuesch/hint-playground/blob/8892a5bcf4207b288d9e8c1baecaf1308bafb0eb/README.adoc#trying-to-build-with-asterius).

the constraint `<0 && >=4.14 && <4.15` seems impossible to satisfy, so this might be the negative (failure) result I was expecting. however I will poke at this a bit more to be sure.
